sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveMetastoreCatalog.scala

```diff
@@ -157,6 +157,14 @@ private[hive] class HiveMetastoreCatalog(hive: HiveContext) extends Catalog with
         tbl.setProperty(s"spark.sql.sources.schema.part.${index}", part)
       }
     }
+
+    // Set the fields for consistent with Hive MetaStore API(.getCols()) accessing
+    val schema = ResolvedDataSource(hive, userSpecifiedSchema, provider, options).relation.schema
+    tbl.setFields(schema.iterator.map { attr =>
+      new FieldSchema(attr.name, toMetastoreType(attr.dataType), null)
```

Comment:
Yes, that's a good question, ideally, we should throw exception if we need to put the unsupported schema (in Hive) into Hive Metastore, right? Otherwise it always break the other application like Hive/Pig when they share the same metastore. That's why I put a `TODO` below, probably we need to provide the Hive Storage Handler for `DataSource` API, it will be great helpful for the legacy systems.